presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1450,7 +1455,7 @@ else if (node.getType() == FULL) {
                 Expression expression = ((JoinOn) criteria).getExpression();
 
                 // need to register coercions in case when join criteria requires coercion (e.g. join on char(1) = char(2))
-                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, output);
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, output, node.getType() == INNER ? Correlation.ALLOWED : Correlation.DISALLOWED);
```

Comment:
it's nice this code line has a (preexisting) comment, but please  add a comment how the processing depends on the join type